src/lib.rs

```diff
@@ -67,21 +68,102 @@ impl Game {
                 0.0,
                 f64::consts::PI * 2.0,
             )
-            .ok()?;
+            .map_err(|_| "Failed to draw an arc.")?;
 
         // Draw the right eye.
         context.move_to(95.0, 65.0);
         context
             .arc(90.0, 65.0, 5.0, 0.0, f64::consts::PI * 2.0)
-            .ok()?;
+            .map_err(|_| "Failed to draw an arc.")?;
 
         context.stroke();
 
-        Some(true)
+        Ok(())
+    }
+
+    // These params describe whether the player is currently holding down each
+    // of the buttons.
+    pub fn draw(&mut self, up: bool, down: bool, left: bool, right: bool) -> Option<String> {
+        match self.draw_internal(up, down, left, right) {
+            Ok(()) => None,
+            Err(e) => Some(format!("Error: {}", e)),
+        }
+    }
+}
+
+struct CanvasInfo {
+    // The 2d rendering context
+    context: web_sys::CanvasRenderingContext2d,
+    // Dimensions of the canvas (for now we assume this won't change)
+    width: f64,
+    height: f64,
+}
+
+fn get_canvas_info() -> Result<CanvasInfo, Box<dyn Error>> {
+    let document = web_sys::window()
+        .ok_or("Could not get window")?
+        .document()
+        .ok_or("Could not get document")?;
```

Comment:
And the award for "Most Verbose Form of Optional Chaining" goes to..

Reply:
lol, true!

having been writing some go lately though, this is still better!